api/bibliotheca.py

```diff
@@ -1217,58 +1187,54 @@ def process_items(self, identifiers):
+
+    def _process_metadata(
+        self, metadata, identifiers_by_bibliotheca_id,
         identifiers_not_mentioned_by_bibliotheca
     ):
-        """Process a single CirculationData object retrieved from
-        Bibliotheca.
+        """Process a single Metadata object (containing CirculationData)
+        retrieved from Bibliotheca.
         """
-        bibliotheca_id = circ[Identifier][Identifier.BIBLIOTHECA_ID]
+        bibliotheca_id = metadata.primary_identifier.identifier
         identifier = identifiers_by_bibliotheca_id[bibliotheca_id]
-        identifiers_not_mentioned_by_bibliotheca.remove(identifier)
-        pools = [lp for lp in identifier.licensed_through
-                 if lp.data_source.name==DataSource.BIBLIOTHECA
-                 and lp.collection == self.collection]
-        if not pools:
-            # We don't have a license pool for this work. That
-            # shouldn't happen--how did we know about the
-            # identifier?--but it shouldn't be a big deal to
-            # create one.
-            pool, ignore = LicensePool.for_foreign_id(
-                self._db, self.collection.data_source, identifier.type,
-                identifier.identifier, collection=self.collection
-            )
-
-            # Bibliotheca books are never open-access.
-            pool.open_access = False
+        if identifier in identifiers_not_mentioned_by_bibliotheca:
+            # Bibliotheca mentioned this identifier. Remove it from
+            # this list so we know the title is still in the collection.
+            identifiers_not_mentioned_by_bibliotheca.remove(identifier)
 
+        edition, is_new = metadata.edition(self._db)
+        pool, is_new = metadata.circulation.license_pool(self._db, self.collection)
+        if is_new:
+            # We didn't have a license pool for this work. That
+            # shouldn't happen--how did we know about the
+            # identifier?--but now we do.
             for library in self.collection.libraries:
                 self.analytics.collect_event(
                     library, pool, CirculationEvent.DISTRIBUTOR_TITLE_ADD,
                     datetime.utcnow()
                 )
-        else:
-            [pool] = pools
-
-        self.api.apply_circulation_information_to_licensepool(
-            circ, pool, self.analytics
-        )
+        edition, changed = metadata.apply(edition, collection=self.collection,
+                                          replace=self.replacement_policy)
+        if changed:
+            # Bibliotheca's metadata for this work has changed.
+            work = pool.work
+            if work:
+                # The work's presentation will be completely
+                # recalculated the next time work_classification runs.
+                WorkCoverageRecord.add_for(
+                    work, WorkCoverageRecord.CLASSIFY_OPERATION,
+                    status=WorkCoverageRecord.REGISTERED
+                )
```

Comment:
This is brand new. I think we should come up with a simple way of doing this, similar to `Work.external_index_needs_updating`. We need to do this every time we get updated metadata from a data source. (I believe with the metadata wrangler we recalculate the presentation immediately.)

Reply:
I filed https://jira.nypl.org/browse/SIMPLY-1485 for this and would like your thoughts.

Reply:
That sounds good, I like the idea of handling this in the metadata layer.